clients/src/main/java/org/apache/kafka/common/network/Authenticator.java

```diff
@@ -54,4 +55,104 @@ default void handleAuthenticationFailure() throws IOException {
      * returns true if authentication is complete otherwise returns false;
      */
     boolean complete();
+
+    /**
+     * Return true if this instance is a client-side authenticator that supports
+     * re-authentication, otherwise false
+     * 
+     * @return true if this instance is a client-side authenticator that supports
+     *         re-authentication, otherwise false
+     */
+    default boolean supportsClientReauth() {
```

Comment:
👍 I eliminated these two methods by adding `Long clientSessionReauthenticationTimeMs()` and `Long serverSessionExpirationTimeNanos()`.  These new methods only ever return non-null on the client and server, respectively.  `clientSessionReauthenticationTimeMs()` now contains the 85%-95% calculation that was in `KafkaChannel` and you asked if it could be moved into the authenticator.